grub-core/loader/multiboot_mbi2.c

```diff
@@ -734,8 +735,77 @@ retrieve_video_parameters (grub_properly_aligned_t **ptrorig)
   return GRUB_ERR_NONE;
 }
 
+void
+grub_multiboot2_setup_slr_table (grub_uint32_t mbi_target, void *mbi_mem)
```

Comment:
Whole MBI is measured by Xen, so there is little reason to measure parts of it again. We probably should just add an entry for MBI (either with implicit size, as it currently is read from MBI itself, or explicitly set in SLRT entry for extra security) and each module, e.g. at the time they are iterated over in `grub_multiboot2_make_mbi()`. In that case, TXT stuff would probably have to be moved from `normal_boot()` to `grub_multiboot_boot()`, which would generate some problems with UEFI path, but we don't care about it (yet) so throwing an error would be enough for now.

Reply:
Can't MBI change while staying the same, say, when pointer values change because allocation code was updated?  Or is a failure in this case expected?

> e.g. at the time they are iterated over in `grub_multiboot2_make_mbi()`

If you want to avoid parsing of MBI, can just iterate over `modules` global variable which shouldn't be freed.

Reply:
> Can't MBI change while staying the same, say, when pointer values change because allocation code was updated?

Fair point, I haven't thought about it. This would require strict description of what isn't measured, and this would probably change with new tags added in the future. On the other hand, such change would most likely make EFI_MMAP different as well, and it is measured.

Reply:
Now that I think about it, SLRT was created by an entity that isn't measured in DRTM chain, so we can't trust it to be valid _until SLRT is measured_. Since addresses aren't measured according to Annex A, there is no guaranty that these addresses are the same as in MBI, so we may end up measuring different region that that which will be later executed. We would need to explicitly check in Xen if those addresses are the same. I wonder if Linux does the same for ramdisk, @dpsmith @rossphilipson any insights?

---

Going back on track, I'd say that MBI should be treated the same way Linux boot params are handled. It serves similar role, also contains pointers, and IIUC it is measured as a whole. One significant difference is that MBI has the command line embedded, while Linux holds just the pointer.

Reply:
Given absence of any active discussion here, should we just do it some way and change things later if it turns out to be inadequate?

> Going back on track, I'd say that MBI should be treated the same way Linux boot params are handled. It serves similar role, also contains pointers, and IIUC it is measured as a whole.

It's measured as a whole and then pointers are followed to measure at least some other parts as well (setup data, command-line, UEFI memmap).

Reply:
https://github.com/TrenchBoot/grub/compare/87f124a95cc8acc45103d9de33df11e8240d1776..94c7f4f9d2f33d3bf503c6356f779261674cf351#diff-285279252527f5bc911013189ac09a0880271f60dc8f28f28fe68c2f5d517c74

Most other changes replace `mbi_mem` with `mbi_size`.